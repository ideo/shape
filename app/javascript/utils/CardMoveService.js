import _ from 'lodash'
import { runInAction } from 'mobx'
import { apiStore as globalApiStore, uiStore as globalUiStore } from '~/stores'

export default class CardMoveService {
  // stores can be passed in e.g. for unit testing, but default to the imported ones
  constructor({ apiStore = globalApiStore, uiStore = globalUiStore } = {}) {
    this.apiStore = apiStore
    this.uiStore = uiStore
  }

  static async moveCards(placement, overrideData = {}) {
    return new this().moveCards(placement, overrideData)
  }

  static moveErrors(opts) {
    return new this().moveErrors(opts)
  }

  async moveCards(placement, overrideData = {}) {
    const { apiStore, uiStore } = this
    const {
      viewingCollection,
      movingFromCollectionId,
      movingCardIds,
      movingIntoCollection,
      cardAction,
      overflowFromMDL,
    } = uiStore

    let data = {
      to_id: viewingCollection ? viewingCollection.id : null,
      from_id: movingFromCollectionId,
      collection_card_ids: movingCardIds,
      placement,
    }
    _.assign(data, overrideData)
    // always ensure non-observable array, esp. when these same ids are used by UndoStore
    data.collection_card_ids = [...data.collection_card_ids]
    // Viewing collection might not be set, such as on the search page
    if (!data.to_id) {
      uiStore.alert("You can't move an item here")
      return
    }

    const toCollection = apiStore.find('collections', data.to_id)
    const movingFromCollection = apiStore.find('collections', data.from_id)

    const error = this.moveErrors({
      toCollection,
      cardAction,
    })

    if (error) {
      if (!toCollection.can_edit_content) {
        uiStore.confirm({
          prompt: error,
          confirmText: 'Continue',
          iconName: 'Alert',
          onConfirm: () => {},
          onCancel: () => {
            uiStore.setMovingCards([])
            uiStore.update('selectedCardIds', [])
          },
        })
      } else {
        uiStore.alert(error)
      }
      return
    }

    const movingWithinCollection =
      !overflowFromMDL &&
      (movingFromCollection === toCollection && cardAction === 'move')
    try {
      uiStore.update('isLoadingMoveAction', true)
      let successMessage
      let res = {}
      switch (cardAction) {
        case 'move':
          if (movingWithinCollection) {
            const order =
              placement === 'beginning'
                ? 0
                : toCollection.collection_card_count + 100
            await this.updateCardsWithinCollection({
              movingIds: data.collection_card_ids,
              collection: toCollection,
              action: 'move',
              placeholder: {
                order,
              },
            })
          } else {
            res = await apiStore.moveCards(data)
          }
          successMessage = 'Items successfully moved!'
          break
        case 'link':
          res = await apiStore.linkCards(data)
          successMessage = 'Items successfully linked!'
          break
        case 'duplicate':
          res = await apiStore.duplicateCards(data)
          successMessage = 'Items successfully duplicated!'
          break
        case 'useTemplate': {
          data = {
            parent_id: data.to_id,
            template_id: data.from_id,
            placement,
          }
          res = await apiStore.createTemplateInstance({
            data,
            template: toCollection,
          })
          successMessage = 'Your template instance has been created!'
          break
        }
        default:
          return
      }
      const meta = res.meta || {}
      // if we received a bulk operation placeholder, place that in the collection
      if (meta.placeholder && toCollection === viewingCollection) {
        runInAction(() => {
          toCollection.collection_cards.unshift(res.data)
        })
      } else if (!movingWithinCollection) {
        // always refresh the current collection
        const fetchData = {}
        if (data.placement && data.placement.row) {
          fetchData.rows = [data.placement.row, data.placement.row + 20]
        }
        await viewingCollection.API_fetchCards(fetchData)
      }

      uiStore.update('isLoadingMoveAction', false)
      uiStore.popupSnackbar({ message: successMessage })
      uiStore.resetSelectionAndBCT()
      uiStore.closeMoveMenu()

      if (!movingIntoCollection) {
        if (placement === 'beginning') {
          uiStore.scrollToTop()
        } else if (placement === 'end') {
          uiStore.scrollToBottom()
        }
      }

      if (
        !meta.placeholder &&
        ((cardAction === 'move' && !movingIntoCollection) ||
          cardAction === 'duplicate' ||
          cardAction === 'link')
      ) {
        if (movingWithinCollection) {
          // reselect moved cards
          uiStore.reselectCardIds(data.collection_card_ids)
        } else if (viewingCollection === toCollection && meta.new_cards) {
          // select newly created cards
          uiStore.reselectCardIds(meta.new_cards)
        }
      }
      return true
    } catch (e) {
      uiStore.update('isLoadingMoveAction', false)
      let message = 'You cannot move a collection within itself.'
      if (e && e.error && e.error[0]) {
        message = e.error[0]
      }
      uiStore.alert(message)
      return false
    }
  }

  updateCardsWithinCollection({
    movingIds,
    collection,
    action,
    placeholder,
    onConfirm = null,
    onCancel = null,
  }) {
    let undoMessage = 'Card move undone'
    const updates = []
    const { apiStore } = this

    const { order, original } = placeholder
    let { width, height } = placeholder

    // don't resize the card for a drag, only for an actual resize
    if (action === 'resize') {
      // just some double-checking validations
      if (height > 2) height = 2
      if (width > 4) width = 4
      // set up action to undo
      if (original.height !== height || original.width !== width) {
        undoMessage = 'Card resize undone'
      }
      updates.push({
        card: original,
        order,
        width,
        height,
      })
    }
    if (movingIds.length > 0) {
      const movingCards = apiStore
        .findAll('collection_cards')
        .filter(card => _.includes(movingIds, card.id))

      // Set order for moved cards so they are between whole integers,
      // and API_batchUpdateCards will properly set/reorder it amongst the collection
      const sortedCards = _.sortBy(movingCards, 'order')
      const toPinAllMovingCards = this.calculateToPinAllMovingCards(
        collection,
        order
      )

      _.each(sortedCards, (card, idx) => {
        const sortedOrder = this.calculateOrderForMovingCard(order, idx)
        const update = {
          card,
          order: sortedOrder,
          pinned: toPinAllMovingCards,
        }
        updates.push(update)
      })
    }

    // Perform batch update on all cards,
    // and show confirmation if this is a template
    return collection.API_batchUpdateCardsWithUndo({
      updates,
      // on a board collection we can just update the moving cards
      updateAllCards: !collection.isBoard,
      undoMessage,
      onConfirm,
      onCancel,
    })
  }

  calculateOrderForMovingCard = (order, index) => {
    return Math.ceil(order) + index
  }

  calculateToPinAllMovingCards = (collection, order) => {
    const { sortedCards } = collection
    const hasPinnedCards = _.some(sortedCards, cc => cc.isPinned)
    if (!hasPinnedCards) return false

    const firstMovingCardSortOrder = this.calculateOrderForMovingCard(order, 0)
    const firstMovingCardIndex = _.findIndex(sortedCards, cc => {
      return cc.order === firstMovingCardSortOrder
    })

    if (firstMovingCardIndex === -1)
      return (_.last(sortedCards) && _.last(sortedCards).isPinned) || false
    if (firstMovingCardIndex <= 1) return true // pin card if moving card in tbe beginning or next to a pinned card
    const leftOfFirstMovingCardIndex = firstMovingCardIndex - 1
    const leftOfFirstMovingCard = sortedCards[leftOfFirstMovingCardIndex]
    // copy pinned state of the card to the left of the first moving card
    return leftOfFirstMovingCard.isPinned
  }

  moveErrors({ toCollection, cardAction }) {
    if (!toCollection.can_edit_content) {
      return 'You only have view access to this collection. Would you like to keep moving the cards?'
    } else if (cardAction === 'useTemplate' && toCollection.isTemplate) {
      return "You can't create a template instance inside another template. You may be intending to create or duplicate a master template into here instead."
    } else if (toCollection.isTestCollection) {
      return "You can't move cards into a test collection"
    }
    return ''
  }
}
