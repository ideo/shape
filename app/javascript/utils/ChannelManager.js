import ActionCableConsumer from './ActionCableConsumer'

class ChannelManager {
  channels = {}

  subscribe(channelName, recordId, callbacks = {}) {
    const channelId = this.channelId(channelName, recordId)
    if (this.channels[channelId]) return this.channels[channelId]
    const channel = ActionCableConsumer.subscriptions.create(
      {
        channel: channelName,
        id: recordId,
      },
      {
        connected: callbacks.channelConnected || function() {},
        disconnected: callbacks.channelDisconnected || function() {},
        received: callbacks.channelReceivedData || function() {},
        rejected: callbacks.channelRejected || function() {},
      }
    )
    this.channels[channelId] = channel
    return channel
  }

  channelId = (channel, recordId = 'home') => `${channel}_${recordId}`

  getChannel(channel, recordId) {
    return this.channels[this.channelId(channel, recordId)]
  }

  unsubscribeAllFromChannel(channelName, { keepOpen = false } = {}) {
    Object.keys(this.channels).forEach(channelId => {
      if (channelId.split('_')[0] === channelName) {
        const channel = this.channels[channelId]
        if (!keepOpen) {
          channel.unsubscribe()
        }
        delete this.channels[channelId]
      }
    })
  }
}

export default new ChannelManager()
