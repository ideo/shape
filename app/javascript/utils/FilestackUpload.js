import axios from 'axios'
import filestack from 'filestack-js'

const API_KEY = process.env.FILESTACK_API_KEY

export const MAX_SIZE = 25 * 1024 * 1024

const imageUploadConfig = {
  accept: ['.pdf', 'image/*', 'application/*', 'text/*', '.docx', '.ppt'],
  maxFiles: 1,
  imageMax: [7500, 7500],
  maxSize: MAX_SIZE,
  transformations: {
    crop: {
      aspectRatio: 5 / 4,
    },
  },
}

const multiImageUploadConfig = {
  ...imageUploadConfig,
  maxFiles: 20,
}

const dropPaneDefaults = {
  overlay: false,
  showIcon: false,
  customText: ' ',
}

class FilestackUpload {
  static get client() {
    return filestack.init(API_KEY)
  }

  static async processFiles(filesUploaded) {
    const filesAttrs = filesUploaded.map(async file => {
      const fileAttrs = {
        handle: file.handle,
        filename: file.filename,
        size: file.size,
        mimetype: file.mimetype,
        url: file.url,
        docInfo: null,
      }
      if (file.mimetype.split('/')[0] === 'image') {
        fileAttrs.url = this.transformedImageUrl(file.handle)
      } else if (file.mimetype === 'application/pdf') {
        const docinfoUrl = this.client.transform(file.handle, {
          output: { docinfo: true },
        })
        const docResp = await axios.get(docinfoUrl)
        fileAttrs.docinfo = docResp.data
      }
      return fileAttrs
    })

    return Promise.all(filesAttrs)
  }

  static async pickImage(opts = {}) {
    return FilestackUpload.pickOneOrMore({ multiple: false, ...opts })
  }

  static async pickImages(opts = {}) {
    return FilestackUpload.pickOneOrMore({ multiple: true, ...opts })
  }

  static async pickOneOrMore({ onSuccess, onFailure, multiple } = {}) {
    const config = multiple ? multiImageUploadConfig : imageUploadConfig
    const resp = await this.client.pick(config)
    if (resp.filesUploaded.length > 0) {
      const filesAttrs = await this.processFiles(resp.filesUploaded)
      if (onSuccess) onSuccess(multiple ? filesAttrs : filesAttrs[0])
    } else if (onFailure) {
      onFailure(resp.filesFailed)
    }
    return resp
  }

  static pdfCoverUrl(handle) {
    return this.client.transform(handle, {
      output: { format: 'png' },
      resize: { fit: 'max', width: 400 },
    })
  }

  static transformedImageUrl(handle) {
    return this.client.transform(handle, {
      rotate: { deg: 'exif' },
    })
  }

  static preview(handle, id) {
    return this.client.preview(handle, { id })
  }

  static makeDropPane(opts = {}, uploadOpts = {}) {
    const config = Object.assign({}, dropPaneDefaults, opts)
    const uploadConfig = Object.assign({}, multiImageUploadConfig, uploadOpts)
    return this.client.makeDropPane(config, uploadConfig)
  }
}

export default FilestackUpload
