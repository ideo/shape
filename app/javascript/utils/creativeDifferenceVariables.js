// Use a map so we can preserve the order of the qualities
export const creativeQualities = new Map()
creativeQualities.set('purpose', {
  color: '#9874AB',
  subqualities: ['usefulness', 'passion', 'clarity'],
})
creativeQualities.set('looking out', {
  color: '#9A4F7A',
  subqualities: [
    'market insightfulness',
    'tech insightfulness',
    'customer insightfulness',
  ],
})
creativeQualities.set('experimentation', {
  color: '#D26A3B',
  subqualities: [
    'user testing',
    'high-fidelity prototyping',
    'low-fidelity prototyping',
    'modeling',
  ],
})
creativeQualities.set('collaboration', {
  color: '#EED950',
  subqualities: [
    'network informality',
    'team collaboration',
    'collaboration characteristics',
  ],
})
creativeQualities.set('empowerment', {
  color: '#83CCB1',
  subqualities: [
    'opposability',
    'fairness',
    'autonomy',
    'risk tolerance',
    'visionary in implementation',
  ],
})
creativeQualities.set('refinement', {
  color: '#8A929D',
  subqualities: ['expert', 'technical creativity', 'detail orientation'],
})

export const allQualityColors = {}
// Iterate over all qualities and subqualities
// and populate object with quality as key and color as value
creativeQualities.forEach((value, key) => {
  allQualityColors[key] = value.color
  value.subqualities.forEach(subquality => {
    allQualityColors[subquality] = value.color
  })
})

export const methodLibraryTagsByType = {
  creativeQualities,
  categories: [
    'systematizing design process',
    'developing and nurturing talent',
    'design process',
    'experience design',
    'building and running labs',
    'developing creative problem solving capabilities',
    'building ventures',
    'research',
    'business models',
    'strategy',
    'marketing',
    'org design',
    'business design',
    'creative work',
  ],
  types: [
    'challenge',
    'worksheet',
    'ritual',
    'process',
    'article',
    'case study',
  ],
}

export const methodLibraryTags = [
  ...Object.keys(allQualityColors),
  ...methodLibraryTagsByType.categories,
  ...methodLibraryTagsByType.types,
]
