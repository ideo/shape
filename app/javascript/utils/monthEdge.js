const monthEdge = (momentDate, timeframe) => {
  const startOfMonth = momentDate.clone().startOf('month')
  const endOfMonth = momentDate.clone().endOf('month')

  const startAllowance = timeframe === 'day' ? 0 : 2
  const endAllowance = timeframe === 'day' ? -1 : 3

  const startDiff = Math.abs(startOfMonth.diff(momentDate, 'days'))
  const endDiff = Math.abs(
    momentDate
      .clone()
      .endOf('month')
      .diff(momentDate, 'days')
  )

  if (startDiff <= startAllowance) {
    // If a date is near the beginning of the month
    // return the previous month
    return startOfMonth.subtract(1, 'month')
  } else if (endDiff <= endAllowance) {
    // If a date is near the end of the month
    // return the end of the month
    return endOfMonth
  }

  return false
}

export default monthEdge
