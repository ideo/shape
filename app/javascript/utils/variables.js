export const ITEM_TYPES = {
  TEXT: 'Item::TextItem',
  FILE: 'Item::FileItem',
  VIDEO: 'Item::VideoItem',
  LINK: 'Item::LinkItem',
  QUESTION: 'Item::QuestionItem',
  CHART: 'Item::ChartItem',
  DATA: 'Item::DataItem',
  EXTERNAL_IMAGE: 'Item::ExternalImageItem',
  LEGEND: 'Item::LegendItem',
}

export const COLLECTION_TYPES = {
  TEST: 'Collection::TestCollection',
  TEST_DESIGN: 'Collection::TestDesign',
}

export const DATASET_CHART_TYPES = {
  AREA: 'area',
  LINE: 'line',
}

export const DATA_MEASURES = [
  { name: 'Participants', value: 'participants' },
  { name: 'Viewers', value: 'viewers' },
  { name: 'Activity', value: 'activity' },
  { name: 'Content Use', value: 'content', tooltip: 'content activities' },
  { name: 'Collections', value: 'collections' },
  { name: 'Items', value: 'items' },
  { name: 'Collections & Items', value: 'records' },
]

export const TEST_COLLECTION_SELECT_OPTIONS = [
  { value: '', label: 'select question type' },
  { value: 'question_context', label: 'Context Setting' },
  { value: 'question_media', label: 'Photo or Video of Idea' },
  { value: 'question_description', label: 'Idea Description' },
  { value: 'question_useful', label: 'Useful' },
  { value: 'question_open', label: 'Open Response' },
  { value: 'question_excitement', label: 'Excitement' },
  { value: 'question_clarity', label: 'Clarity' },
  { value: 'question_different', label: 'Different' },
  { value: 'question_category_satisfaction', label: 'Category Satisfaction' },
]

export const KEYS = {
  ENTER: 13,
  ESC: 27,
}

export default {
  headerHeight: 50,
  maxWidth: 1320,
  containerPadding: {
    horizontal: 2,
  },
  topScrollTrigger: 210,
  maxTitleLength: 144,
  actionMenuWidth: 250,

  // Keep in sync with assets/stylesheets/core/base.scss
  responsive: {
    // based on going from 2 -> 1 column
    smallBreakpoint: 645,
    // based on going from 3 -> 2 cols
    medBreakpoint: 1060,
    largeBreakpoint: 1308,
    // material-ui Grid `sm` == 960px
    muiSmBreakpoint: 960,
  },

  transition: 'all 0.2s',
  transitionWithDelay: 'all 0.3s 0.2s',

  navOpacity: 0.9,

  cardHoldTime: 0.4 * 1000,

  colors: {
    black: '#120f0e',
    white: '#ffffff',
    commonLightest: '#f5f4f3',
    commonLight: '#f2f1ee',
    commonMedium: '#c6c1bf',
    commonDark: '#a89f9b',
    commonDarkest: '#787878',
    primaryLightest: '#f0f4f6',
    primaryLight: '#c0dbde',
    primaryMedium: '#9ec1cc',
    primaryMediumDark: '#86b2b7',
    primaryDark: '#5698ae',
    primaryDarkest: '#35889e',
    secondaryLight: '#73808f',
    secondaryMedium: '#5a6a7c',
    secondaryDark: '#3f526a',
    secondaryDarkest: '#242d37',
    tertiaryMedium: '#de8f74',
    tertiaryDark: '#c43a31',
    ctaPrimary: '#00a0e0',
    alert: '#d66742',
    caution: '#fcf113',
    offset: '#6a7c7e',
    placeholderGray: '#bcbcbc',
    transparent: 'transparent',
  },

  iconSizes: {
    bct: 47,
  },

  fonts: {
    baseSize: 16,
    sans: "'Gotham', sans-serif",
    serif: "'Sentinel', serif",
  },

  weights: {
    book: 400,
    medium: 500,
    bold: 700,
  },

  zIndex: {
    aboveClickWrapper: 501,
    clickWrapper: 500,
    activityLog: 251,
    // NOTE: if globalHeader is > pageHeader
    // then it will also be above the EditableName ClickWrapper
    cardDragging: 222,
    globalHeader: 221,
    pageHeader: 220,
    popoutMenu: 201,
    scrollIndicator: 200,
    cardHovering: 160,
    gridCard: 150,
    gridCardTop: 151,
    commentMentions: 100,
    commentHeader: 10,
    itemClose: 10,
    floatOverContent: 2,
    gridCardBg: 1,
    aboveVictoryChart: 100,
  },

  quillDefaults: {
    formats: ['link', 'header'],
    modules: {
      toolbar: [
        // header: false means "normal" i.e. no formatting
        [{ header: [3, false] }],
        ['link'],
      ],
    },
  },

  defaults: {
    video: {
      name: 'Video',
    },
  },
}
