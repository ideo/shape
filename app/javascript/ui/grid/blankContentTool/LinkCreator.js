import _ from 'lodash'
import PropTypes from 'prop-types'

import VideoUrl from '~/utils/VideoUrl'
import parseURLMeta from '~/utils/parseURLMeta'
import { ITEM_TYPES } from '~/utils/variables'
import GenericLinkCreator from './GenericLinkCreator'

class LinkCreator extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      url: '',
      urlValid: false,
      loading: false,
      name: '',
      thumbnailUrl: '',
      meta: {},
      passwordRequired: false,
      password: '',
    }
    this.canceled = false
    this.parseMetadata = _.debounce(this._parseMetadata, 500)
    this.lookupVideoAPI = _.debounce(this._lookupVideoAPI, 500)
  }

  componentWillUnmount() {
    this.canceled = true
  }

  onUrlChange = e => {
    const url = e.target.value
    const { type } = this.props
    this.setState({
      url,
      loading: true,
    })
    if (VideoUrl.isValid(url)) {
      return this.lookupVideoAPI(url)
    }
    if (type === 'video') {
      // Video does not accept normal generic links
      this.setState({
        urlValid: false,
        loading: false,
      })
      return false
    }
    return this.parseMetadata(url)
  }

  onPasswordChange = e => {
    const password = e.target.value
    const { url } = this.state
    this.setState({
      loading: true,
      password,
    })
    if (VideoUrl.isValid(url)) {
      return this.lookupVideoAPI(url, password)
    }
    this.setState({
      urlValid: false,
      loading: false,
    })
    return false
  }

  _parseMetadata = async url => {
    if (url.length <= 3) return
    const meta = await parseURLMeta(url)
    if (this.canceled) return
    this.setState({ loading: false })
    if (meta && (meta.title || meta.shapeLink)) {
      this.setState({
        meta,
        urlValid: 'link',
      })
    } else {
      this.setState({ urlValid: false })
    }
  }

  _lookupVideoAPI = async (url, password = '') => {
    const {
      name,
      thumbnailUrl,
      passwordRequired,
    } = await VideoUrl.getAPIdetails(url, password)
    if (this.canceled) return
    this.setState({ loading: false })
    if (name) {
      this.setState({ name, thumbnailUrl, urlValid: 'video', passwordRequired })
    } else {
      this.setState({ urlValid: false })
    }
  }

  createItem = e => {
    e.preventDefault()
    const { urlValid } = this.state
    if (!urlValid) return
    if (urlValid === 'link') {
      this.createLinkItem()
    } else if (urlValid === 'video') {
      this.createVideoItem()
    }
  }

  createVideoItem = () => {
    const { url, name, thumbnailUrl } = this.state
    // Get a normalized URL to make it easier to handle in our system
    const { normalizedUrl } = VideoUrl.parse(url)
    const attrs = {
      item_attributes: {
        type: ITEM_TYPES.VIDEO,
        url: normalizedUrl,
        name,
        thumbnail_url: thumbnailUrl,
      },
    }
    this.props.createCard(attrs)
  }

  createLinkItem = () => {
    const { url, meta } = this.state
    let attrs = {
      item_attributes: {
        type: ITEM_TYPES.LINK,
        url,
        name: meta.title,
        content: meta.description,
        thumbnail_url: meta.image,
        icon_url: meta.icon,
      },
    }
    if (meta.shapeLink) {
      attrs = {
        type: 'CollectionCard::Link',
      }
      attrs[`${meta.recordType}_id`] = meta.recordId
    }
    this.props.createCard(attrs)
  }

  render() {
    const { type } = this.props
    const { url, urlValid, passwordRequired, password } = this.state
    const loading = this.props.loading || this.state.loading

    return (
      <GenericLinkCreator
        url={url}
        urlValid={!!urlValid}
        placeholder={`${_.capitalize(type)} URL`}
        onSubmit={this.createItem}
        onChange={this.onUrlChange}
        onClose={this.props.closeBlankContentTool}
        loading={loading}
        password={password}
        passwordField={passwordRequired}
        onPasswordChange={this.onPasswordChange}
      />
    )
  }
}

LinkCreator.propTypes = {
  type: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
  createCard: PropTypes.func.isRequired,
  closeBlankContentTool: PropTypes.func.isRequired,
}

export default LinkCreator
