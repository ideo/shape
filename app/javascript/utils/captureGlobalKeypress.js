import _ from 'lodash'
import { apiStore, uiStore, undoStore } from '~/stores'
import { quillSelectors } from '~/utils/variables'

const quillEditorClick = e => {
  return e.target.closest(quillSelectors)
}

const emptySpaceClick = e => {
  const { target } = e
  return !!(
    target.getAttribute && target.getAttribute('data-empty-space-click')
  )
}

export const handleMouseDownSelection = e => {
  const emptySpaceMouseDown = emptySpaceClick(e)
  const { isEditingText } = uiStore
  const outsideQuillMouseDown = !quillEditorClick(e) && isEditingText
  if (emptySpaceMouseDown) {
    if (outsideQuillMouseDown) {
      uiStore.setCommentingOnRecord(null)
      // we need a tiny delay so that the highlight gets properly unset
      // before kicking out of the textEditingItem
      setTimeout(() => {
        uiStore.update('textEditingItem', null)
      })
    }
    // if we clicked an empty space...
    uiStore.deselectCards()
    uiStore.onEmptySpaceClick(e)
    uiStore.closeBlankContentTool()
    uiStore.closeCardMenu()
    uiStore.update('editingCardCover', null)
    return 'emptySpace'
  }
  return false
}

const captureGlobalKeypress = e => {
  const { activeElement } = document

  // none of the following card editing actions apply when you are editing a test
  if (_.get(uiStore, 'viewingCollection.isTestCollection')) {
    return false
  }
  // allow normal keypress on input element, quill, and draftjs
  const shouldNormalKeyPressBeAllowed =
    activeElement.nodeName === 'INPUT' ||
    _.intersection(activeElement.classList, [
      'ql-editor',
      'ql-container',
      'public-DraftEditor-content',
      'edit-cover-title',
      'edit-cover-subtitle',
    ]).length > 0

  if (shouldNormalKeyPressBeAllowed) return false
  const { code, metaKey, ctrlKey, shiftKey } = e
  const { selectedCardIds, viewingCollection } = uiStore
  const ctrlKeypress = metaKey || ctrlKey
  let card
  switch (code) {
    // CTRL+X: Move
    case 'KeyX':
      if (!viewingCollection || !ctrlKeypress) {
        return false
      }
      if (!selectedCardIds.length) {
        return false
      }
      card = apiStore.find('collection_cards', selectedCardIds[0])
      if (card) {
        // TODO: investigate... shouldn't need to reselect on move, but
        // what if you select pinned cards? API will reject the move... ?
        card.reselectOnlyMovableCards(selectedCardIds)
      }
      viewingCollection.confirmEdit({
        onConfirm: () => {
          uiStore.openMoveMenu({
            from: viewingCollection,
            cardAction: 'move',
          })
        },
      })
      break
    // CTRL+C: Duplicate
    case 'KeyC':
      if (!viewingCollection || !ctrlKeypress) {
        return false
      }
      uiStore.openMoveMenu({
        from: viewingCollection,
        cardAction: 'duplicate',
      })
      break
    // CTRL+V: Place (Paste)
    case 'KeyV':
      if (ctrlKeypress) {
        // MoveModal will listen to this value and then set it to false
        uiStore.update('pastingCards', true)
      }
      break
    case 'KeyZ':
      if (!ctrlKeypress) {
        return false
      }
      // CTRL+Shift+Z: Redo
      if (shiftKey) {
        undoStore.handleRedoKeyPress()
        break
      }
      // CTRL+Z: Undo
      undoStore.handleUndoKeypress()
      break
    // CTRL+A: Select All
    case 'KeyA':
      if (ctrlKeypress && uiStore.viewingCollection) {
        e.preventDefault()
        uiStore.selectAll({ location: 'Global' })
      }
      break
    case 'Backspace':
    case 'Delete':
      if (!selectedCardIds || !selectedCardIds.length) {
        return false
      }
      card = apiStore.find('collection_cards', selectedCardIds[0])
      // see note in CollectionCard model -- this could really be a static method;
      // because it's not, we just have to call it on any selected card
      card.API_archive()
      break
    case 'Escape':
      // save on esc happens only when user clicks the title textarea
      const { editingCardCover } = uiStore
      if (editingCardCover) {
        uiStore.update('editingCardCover', null)
      }
      break
    default:
      break
  }
  return false
}

export default captureGlobalKeypress
