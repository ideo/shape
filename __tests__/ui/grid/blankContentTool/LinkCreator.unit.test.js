import LinkCreator from '~/ui/grid/blankContentTool/LinkCreator'
import expectTreeToMatchSnapshot from '#/helpers/expectTreeToMatchSnapshot'
import { ITEM_TYPES } from '~/utils/variables'

const e = { preventDefault: jest.fn() }
let wrapper, props, component
describe('LinkCreator', () => {
  beforeEach(() => {
    props = {
      type: 'link',
      loading: false,
      createCard: jest.fn(),
      closeBlankContentTool: jest.fn(),
    }
    props.createCard.mockClear()
    wrapper = shallow(<LinkCreator {...props} />)
    component = wrapper.instance()
  })

  it('renders snapshot', () => {
    expectTreeToMatchSnapshot(wrapper)
  })

  it('renders a GenericLinkCreator', () => {
    expect(wrapper.find('GenericLinkCreator').exists()).toBeTruthy()
  })

  describe('with video url', () => {
    it('calls createCard with video data when url is valid', () => {
      component.state = {
        url: 'https://www.youtube.com/watch?v=zDB3NvF9LSI',
        urlValid: 'video',
        name: 'Youtube Video',
        thumbnailUrl: 'http://youtube.com/thumb',
      }
      component.createItem(e)
      expect(props.createCard).toHaveBeenCalledWith({
        item_attributes: {
          type: ITEM_TYPES.VIDEO,
          url: component.state.url,
          name: component.state.name,
          thumbnail_url: component.state.thumbnailUrl,
        },
        filter: 'nothing',
      })
    })

    it('does not call createCard when link is invalid', () => {
      component.state = {
        url: 'httpmy.url.com',
        urlValid: false,
      }
      component.createItem(e)
      expect(props.createCard).not.toHaveBeenCalled()
    })
  })

  describe('with video password props', () => {
    it('sends password props to GenericLinkCreator', () => {
      component.setState({
        passwordRequired: true,
        password: '1234',
      })
      wrapper.update()

      expect(wrapper.find('GenericLinkCreator').props().password).toEqual(
        '1234'
      )
      expect(
        wrapper.find('GenericLinkCreator').props().passwordField
      ).toBeTruthy()
    })
  })

  describe('with generic link url', () => {
    it('calls createCard with link metadata when link is valid', () => {
      const meta = {
        title: 'My Site',
        content: 'Site content here',
        image: 'http://image.url',
        icon: 'http://image.url/icon',
      }
      component.state = {
        urlValid: 'link',
        url: 'http://my.url.com',
        meta,
      }
      component.createItem(e)
      expect(props.createCard).toHaveBeenCalledWith({
        item_attributes: {
          type: ITEM_TYPES.LINK,
          url: component.state.url,
          name: meta.title,
          content: meta.description,
          thumbnail_url: meta.image,
          icon_url: meta.icon,
        },
      })
    })

    it('does not call createCard when link is invalid', () => {
      component.state = {
        urlValid: false,
        url: 'httpmy.url.com',
        meta: {},
      }
      component.createItem(e)
      expect(props.createCard).not.toHaveBeenCalled()
    })
  })
})
