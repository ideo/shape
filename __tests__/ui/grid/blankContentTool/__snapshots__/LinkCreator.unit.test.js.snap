// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkCreator renders snapshot 1`] = `
<GenericLinkCreator
  loading={false}
  onChange={[Function]}
  onClose={[MockFunction]}
  onPasswordChange={[Function]}
  onSubmit={[Function]}
  password=""
  passwordField={false}
  placeholder="Link URL"
  url=""
  urlValid={false}
/>
`;
